File: backend/app/schemas/risk.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


RiskLevel = Literal["low", "medium", "high", "critical"]

Decision = Literal[
    "ALLOW",
    "STEP_UP",
    "MANUAL_REVIEW",
    "BLOCK",
]


class SignalScores(BaseModel):
    acoustic: float = Field(ge=0.0, le=1.0)
    prosody: float = Field(ge=0.0, le=1.0)
    speaker: float = Field(ge=0.0, le=1.0)
    context: float = Field(ge=0.0, le=1.0)
    transaction: float = Field(ge=0.0, le=1.0)
    behaviour: float = Field(ge=0.0, le=1.0)


class RiskFactor(BaseModel):
    name: str
    category: str
    contribution: float
    evidence: str


class RiskResult(BaseModel):
    risk_score: float = Field(ge=0.0, le=1.0)
    risk_level: RiskLevel
    decision: Decision

    signals: SignalScores

    factors: list[RiskFactor] = Field(default_factory=list)

    model_version: str
    calibrated: bool = False

    latency_ms: float | None = None

File: frontend/src/lib/types/risk.ts
export type RiskLevel = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";
export type RiskDecision =
  | "ALLOW"
  | "STEP_UP_VERIFICATION"
  | "MANUAL_REVIEW"
  | "ESCALATE";

export type SignalKey =
  | "acoustic"
  | "prosody"
  | "speaker"
  | "context"
  | "behaviour"
  | "transaction";

export interface RiskSignal {
  key: SignalKey;
  label: string;
  score: number | null;
  summary: string;
  limitation?: string;
}

export interface RiskEvent {
  id: string;
  sessionId: string;
  investigationId?: string;
  title: string;
  summary: string;
  riskScore: number;
  riskLevel: RiskLevel;
  decision: RiskDecision;
  occurredAt: string;
  amount?: string;
  signals: RiskSignal[];
  explanations: string[];
  unavailableEvidence: string[];
  modelVersion: string;
}

export interface Investigation {
  id: string;
  title: string;
  riskLevel: RiskLevel;
  status: "Open" | "Reviewing" | "Resolved";
  recommendedAction: string;
  lastActivity: string;
  eventIds: string[];
  summary: string;
  timeline: Array<{ time: string; title: string; detail: string }>;
}

export interface AnalysisResult {
  risk_score: number;
  risk_level: RiskLevel;
  decision: RiskDecision;
  signals: Record<SignalKey, number>;
  explanations: string[];
  model_version: string;
  latency_ms: number;
}
